contracts/pcv/sentinel/PCVSentinel.sol

```diff
@@ -0,0 +1,173 @@
+    }
+
+    /**
+     * @notice returns a list of all guards
+     */
+    function allGuards() 
+        external
+        view
+        override
+        returns (address[] memory all) 
+    {
+        all = new address[](guards.length());
+
+        for (uint i = 0; i < guards.length(); i++) {
+            all[i] = guards.at(i);
+        }
+
+        return all;
+    }
+
+    /**
+     * @notice returns the list of all guards that pass their check
+     * @dev you *can* use this on-chain if you want, but this is made
+     * to be called off-chain to pass into protecManys and not hyper-efficient
+     */
+    function checkAll()
+        external 
+        view
+        override 
+        returns (address[] memory) 
+    {
+        uint256 numGuardsCanProtec;
+
+        for (uint256 i=0; i<guards.length(); i++) {
+            if(IGuard(guards.at(i)).check()) {
+                numGuardsCanProtec++;
+            }
+        }
+
+        address[] memory guardsWhoCanProtec = new address[](numGuardsCanProtec);
+
+        for (uint256 i=0; i<guards.length(); i++) {
+            if(IGuard(guards.at(i)).check()) {
+                numGuardsCanProtec--;
+                guardsWhoCanProtec[numGuardsCanProtec] = guards.at(i);
+            }
+        }
+
+        return guardsWhoCanProtec;
+    }
+
+    // ---------- Governor-or-Admin-Or-Guardian-Only State-Changing API ----------
+
+    /**
+     * @notice adds a guard contract to the PCV Sentinel
+     * @param guard the guard-contract to add
+     */
+    function knight(address guard) 
+        external 
+        override 
+        isGovernorOrGuardianOrAdmin
+    {
+        guards.add(guard);
+
+        // Inform the kingdom of this glorious news
+        emit GuardAdded(guard);
+    }
+
+    /**
+     * @notice removes a guard
+     * @param traitor the guard-contract to remove
+     */
+    function slay(address traitor) 
+        external 
+        override 
+        isGovernorOrGuardianOrAdmin
+    {
+        guards.remove(traitor);
+
+        // Inform the kingdom of this sudden and inevitable betrayal
+        emit GuardRemoved(traitor);
+    }
+
+    // ---------- Public State-Changing API ----------
+
+    /**
+     * @notice checks the guard and runs its protec actions if needed
+     * @param guard the guard to activate
+     */
+    function protec(address guard) 
+        external
+        override
+    {
+        require(guards.contains(guard), "Guard does not exist.");
+
+        bool needsProtec = IGuard(guard).check();
+
+        if (needsProtec) {
+            (address[] memory targets, bytes[] memory datas) = IGuard(guard).getProtecActions();
+
+            for(uint256 i=0; i<targets.length; i++) {
+                require(targets[i] != address(this), "Nyeh!");
+                targets[i].call(datas[i]); // Unopinionated about reverts
+            }
+
+            emit Protected(guard);
+        } else {
+            revert("No need to protec.");
```

Comment:
no need to revert if there is nothing to protec

instead, have the function return true or false and return true if anything was protected and false otherwise

Reply:
Reverting lets us bubble up errors and emit events when executing multiple protec actions

Reply:
I'm not convinced this is the right decision.

If you pass in an array of guards to protec, and only half of them can be protec'd, why should we revert when the revert is ignored?

Also, any errors that try to bubble up get swallowed and are forgotten so this error string is literally pointless unless you try to call protec individually, and even then, it's much nicer to return a boolean than fail loudly.

If protecMany reverted on errors to protec, then I would be fine with this pattern.